File: backend/app/models/goods.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import BigInteger, DateTime, Index, Integer, Numeric, SmallInteger, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class Goods(Base):
    __tablename__ = "goods"
    __table_args__ = (
        Index("ux_goods_platform_url_hash", "platform", "goods_url_hash", unique=True),
        Index("ix_goods_platform_sale_status", "platform", "sale_status"),
    )

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True, comment="goods id")
    goods_name: Mapped[str] = mapped_column(String(255), nullable=False, comment="goods name")
    brand: Mapped[str] = mapped_column(String(64), default="", comment="brand")
    model: Mapped[str] = mapped_column(String(128), default="", comment="model")
    spec_param: Mapped[str] = mapped_column(String(255), default="", comment="specification")
    platform: Mapped[str] = mapped_column(String(32), nullable=False, comment="platform")
    shop_name: Mapped[str] = mapped_column(String(128), default="", comment="shop name")
    original_price: Mapped[Decimal] = mapped_column(Numeric(10, 2), default=0.00, comment="original price")
    real_price: Mapped[Decimal] = mapped_column(Numeric(10, 2), default=0.00, comment="final price")
    coupon_desc: Mapped[str] = mapped_column(String(512), default="", comment="coupon description")
    goods_img: Mapped[str] = mapped_column(String(512), default="", comment="goods image url")
    goods_url: Mapped[str] = mapped_column(String(800), nullable=False, comment="source goods url")
    goods_url_hash: Mapped[str] = mapped_column(String(64), nullable=False, default="", comment="sha256 of goods url")
    sales_num: Mapped[int] = mapped_column(Integer, default=0, comment="sales count")
    score: Mapped[Decimal] = mapped_column(Numeric(3, 1), default=0.0, comment="rating score")
    feature_content: Mapped[str] = mapped_column(Text, default="", comment="embedding feature text")
    sale_status: Mapped[int] = mapped_column(SmallInteger, default=1, comment="1 on sale, 0 off shelf")
    last_update_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), comment="last updated at")
    create_time: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), comment="created at")

File: frontend/lib/shared/models/goods.dart
class Goods {
  final int id;
  final String goodsName;
  final String brand;
  final String model;
  final String specParam;
  final String platform;
  final String shopName;
  final double originalPrice;
  final double realPrice;
  final String couponDesc;
  final String goodsImg;
  final String goodsUrl;
  final int salesNum;
  final double score;
  final int saleStatus;
  final double matchScore;
  final double trustScore;

  const Goods({
    required this.id,
    required this.goodsName,
    this.brand = '',
    this.model = '',
    this.specParam = '',
    required this.platform,
    this.shopName = '',
    this.originalPrice = 0,
    required this.realPrice,
    this.couponDesc = '',
    this.goodsImg = '',
    required this.goodsUrl,
    this.salesNum = 0,
    this.score = 0,
    this.saleStatus = 1,
    this.matchScore = 0,
    this.trustScore = 0,
  });

  bool get isOnSale => saleStatus == 1;
  bool get isOffShelf => saleStatus == 0;

  String get platformLabel {
    switch (platform) {
      case 'jingdong':
      case '京东':
        return '京东';
      case 'taobao':
      case '淘宝':
        return '淘宝';
      case 'pdd':
      case '拼多多':
        return '拼多多';
      case 'douyin':
      case '抖音':
        return '抖音';
      default:
        return platform;
    }
  }

  String get salesText {
    if (salesNum >= 10000) return '${(salesNum / 10000).toStringAsFixed(1)}万+';
    if (salesNum >= 1000) return '${(salesNum / 1000).toStringAsFixed(1)}k+';
    return '$salesNum';
  }

  String get trustLabel {
    if (trustScore >= 80) return '高可信';
    if (trustScore >= 60) return '中可信';
    return '需核验';
  }

  factory Goods.fromJson(Map<String, dynamic> json) {
    return Goods(
      id: json['goods_id'] as int? ?? 0,
      goodsName: json['goods_name'] as String? ?? '',
      brand: json['brand'] as String? ?? '',
      model: json['model'] as String? ?? '',
      specParam: json['spec_param'] as String? ?? '',
      platform: json['platform'] as String? ?? '',
      shopName: json['shop_name'] as String? ?? '',
      originalPrice: (json['original_price'] as num?)?.toDouble() ?? 0,
      realPrice: (json['real_price'] as num?)?.toDouble() ?? 0,
      couponDesc: json['coupon_desc'] as String? ?? '',
      goodsImg: json['goods_img'] as String? ?? '',
      goodsUrl: json['goods_url'] as String? ?? '',
      salesNum: json['sales_num'] as int? ?? 0,
      score: (json['score'] as num?)?.toDouble() ?? 0,
      saleStatus: json['sale_status'] as int? ?? 1,
      matchScore: (json['match_score'] as num?)?.toDouble() ?? 0,
      trustScore: (json['trust_score'] as num?)?.toDouble() ?? 0,
    );
  }
}
